### ISSUE TITLE
Does defining init= make sense for an extern record?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#15881]
Related to #7992 and #9898.

As part of work I am doing on improving LLVM ABI support, I am running into problems related to the ReturnByRef transformation applying when an `export proc` returns an `extern record`.

Currently, the compiler doesn't produce code with correct memory management unless the ReturnByRef transformation is used. However, in the context of an `export proc`, I think it would be reasonable to expect a returned C `struct` (viewed as an `extern record`) would be returned in the usual manner according to the ABI (e.g. possibly in a register if the struct is small).

In the near term, I'm looking at disabling all memory management for extern records - in effect requiring them to be POD.

However, some comments in previous discussions such as https://github.com/chapel-lang/chapel/issues/7992#issuecomment-587652644 or https://github.com/chapel-lang/chapel/issues/9898#issuecomment-398558256 lead me to think that the approach here needs discussion.

So, the question is: which, if any, of the following methods should be defineable for an `extern record` ?
 * a default initializer such as `init()`
 * a custom initializer accepting more arguments e.g. `init(arg:real)`
 * a copy initializer `init=()`
 * a deinitializer `deinit()`

**Comment by bradcray** [comment]
Capturing some points of discussion with Michael:

* since `extern` things currently are defined as being C-like and C doesn't really have default initializers, copy initializers, etc. I think the "assume/require POD" approach is reasonable for the time being (and will be curious whether we currently have code or tests that break under that assumption/restriction).
* ultimately, I think it would be nice to let a user "beef up" an extern record a bit such that, within Chapel, they can be more powerful and behave more like a Chapel record with default values, etc.  But I think that can come when we're ready for it without breaking code developed under the previous bullet.

To this question:

> If init= can be defined, how should an extern proc that returns a value of that type, or accepts a value of that type with an in intent formal work?

Given:

```chapel
extern proc foo(in x: R);  // where R is a record
```

I think of this as defining the C routine:

```c
void foo(R x) { ... }
```

so my intuition would be to not expect Chapel to call the copy initializer in passing a variable of type `R` to the routine, but to have the call to C effectively do the copy and therefore not call the `init=` routine.  Thus, the `init=` would only be used for copy initializations done within Chapel itself (e.g., `var myR2 = myR;`).

Similarly, for:

```chapel
extern proc bar(): R;
```

I think of that as corresponding to the C routine:

```c
R bar() { ... }
```

So if I said:

```chapel
var myR = bar();
```

I wouldn't imagine calling the copy initializer due to bar() returning an `R` value, but I would call it when copying from bar()'s returned value to `myR`.

How does that fit with your thinking?

**Comment by mppf** [comment]
> To this question:
> 
> > If init= can be defined, how should an extern proc that returns a value of that type, or accepts a value of that type with an in intent formal work?
> 
> Given:
> 
> ```chapel
> extern proc foo(in x: R);  // where R is a record
> ```
> 
> I think of this as defining the C routine:
> 
> ```c
> void foo(R x) { ... }
> ```
> 
> so my intuition would be to not expect Chapel to call the copy initializer in passing a variable of type `R` to the routine, but to have the call to C effectively do the copy and therefore not call the `init=` routine.  Thus, the `init=` would only be used for copy initializations done within Chapel itself (e.g., `var myR2 = myR;`).

In this case, I think it'd be reasonable for the Chapel compiler to create the copy-initialization (if necessary according to the language rules). But for the C function signature to be `void foo(R x) { ... }`, the result of copy-initialization would be bit-copied into the C function argument. However that is OK (since we allow Chapel records to be "moved" around). The C function however would bear the responsibility for calling `deinit` (or C code that does the same thing) before returning - or else generally speaking there would be a memory leak.

I think this approach could work and would make the `init=` more useful but I am not certain there isn't an issue with it I'm not thinking of...

> 
> Similarly, for:
> 
> ```chapel
> extern proc bar(): R;
> ```
> 
> I think of that as corresponding to the C routine:
> 
> ```c
> R bar() { ... }
> ```
> 
> So if I said:
> 
> ```chapel
> var myR = bar();
> ```
> 
> I wouldn't imagine calling the copy initializer due to bar() returning an `R` value, but I would call it when copying from bar()'s returned value to `myR`.
> 
> How does that fit with your thinking?

My expectation would be that `bar` would return a fully-formed `R` and that the statement `var myR = bar()` would use normal Chapel rules and not run the copy initializer. Instead, the returned value would "move" into `myR`.

However this would include an assumption that the `bar()` function is returning a fully-formed `R` and not one that is (for example) aliasing other values that are about to be freed.

It appears that my thinking was different from yours in both cases.

**Comment by mppf** [comment]
> In the near term, I'm looking at disabling all memory management for extern records - in effect requiring them to be POD.

> * since `extern` things currently are defined as being C-like and C doesn't really have default initializers, copy initializers, etc. I think the "assume/require POD" approach is reasonable for the time being (and will be curious whether we currently have code or tests that break under that assumption/restriction).

Just elaborating - defining `init=` or `deinit` on an `extern record` will cause a compilation error under this approach.

**Comment by lydia-duncan** [comment]
I'm a little leery of defining a method for an extern record in such a way that it won't be available to the C side for something as intrinsic to the type's functionality.  It seems like it would lead to confusion, or potentially breaking the type's assumptions about itself, especially depending on what is written in these methods.

**Comment by mppf** [comment]
If it's an extern record, presumably these functionalities already exist on the C side (possibly with explicit functions like `my_c_struct_copy(my_c_struct)` and `my_c_struct_free(my_c_struct)` ? In that event, the expected use case would be that the Chapel `init=` and `deinit` would just be wired up to call these automatically. It's not a problem that the Chapel `init=` and `deinit` don't run in the C code because C programmers will already have the explicit API to do the same thing (with my_c_struct_copy / my_c_struct_free in the example).

**Comment by lydia-duncan** [comment]
That's fair, but I'd feel more comfortable if there ends up being a warning in the documentation describing how to use the feature

**Comment by mppf** [comment]
> since extern things currently are defined as being C-like and C doesn't really have default initializers, copy initializers, etc. I think the "assume/require POD" approach is reasonable for the time being (and will be curious whether we currently have code or tests that break under that assumption/restriction).

I saw 0 failures from the addition of this compile-time checking (i.e. no existing tests issues errors). See PR #15903. Almost all test changes in that PR are from requiring intents - with the exception of test/functions/gbt/retRecordMultiDef.chpl which changed due to the requirement that `export` procs only use `extern` records.

**Comment by mppf** [comment]
PR #18542 zero-initializes variables of extern type but in the future we might expect a `proc init()`, if provided, to be called. Should we make providing a `proc init()` for an extern type be an error for now? What about `deinit` and `init=` ?

**Comment by mppf** [comment]
Right, I think the challenge here has to do with the issues I was bringing up in the top post of this issue:

> Currently, the compiler doesn't produce code with correct memory management unless the ReturnByRef transformation is used. However, in the context of an `export proc`, I think it would be reasonable to expect a returned C `struct` (viewed as an `extern record`) would be returned in the usual manner according to the ABI (e.g. possibly in a register if the struct is small). These two are in direct conflict.

I don't think we should try to fix this now as I think it'd be a bigger effort than is called for in order to have this feature today and additionally I think it makes sense to put it off as we are working on migrating resolution to compiler/next.

We could consider supporting just `init()` for extern records but I am uncertain of how useful that would be without `deinit` and `init=` as well. In particular I worry that somebody finding that they can write `init()` on an extern record will assume the others are available. I don't think we can get to correct code for the latter two without a bigger effort.
